File: huannv_Rx_Retrofit_MultiThread/app/src/main/java/com/example/base_project/upcoming/adapter/ListMovieHorizontalAdapter.java
package com.example.base_project.upcoming.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.example.base_project.R;
import com.example.base_project.main.model.Movie;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

import static com.example.base_project.Utils.KEY_IMAGE;

public class ListMovieHorizontalAdapter extends RecyclerView.Adapter<ListMovieHorizontalAdapter.ViewHolder> {
    private List<Movie> listMovie = new ArrayList<>();
    private Context context;
    private ClickItemListener clickItemListener;

    public ListMovieHorizontalAdapter(Context context, List<Movie> listMovie, ClickItemListener clickItemListener) {
        this.listMovie = listMovie;
        this.context = context;
        this.clickItemListener = clickItemListener;
    }

    public void setListMovie(List<Movie> listMovie) {
        this.listMovie = listMovie;
        notifyDataSetChanged();
    }

    public List<Movie> getListMovie() {
        return listMovie;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.item_movie, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Movie item = listMovie.get(position);
        holder.bindData(item);
    }

    @Override
    public int getItemCount() {
        return listMovie == null ? 0 : listMovie.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder{
        @BindView(R.id.image_movie)
        ImageView movieImageView;
        @BindView(R.id.tv_movie_name)
        TextView movieNameTextView;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (clickItemListener != null) {
                        clickItemListener.onClickMovie(getAdapterPosition());
                    }
                }
            });
        }

        public void bindData(Movie item) {
            String urlImage = KEY_IMAGE + item.getPosterPath();
            Glide.with(movieImageView)
                    .load(urlImage)
                    .into(movieImageView);
            movieNameTextView.setText(item.getTitle());

        }
    }

    public interface ClickItemListener {
        void onClickMovie (int position);
    }
}

File: Trainning_Kotlin/app/src/main/java/com/example/myapplication/main/adapter/ListMovieHorizontalAdapter.kt
package com.example.myapplication.main.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import com.bumptech.glide.Glide
import com.example.myapplication.R
import com.example.myapplication.Utils
import com.example.myapplication.Utils.Companion.KEY_IMAGE
import com.example.myapplication.model.Movie

class ListMovieHorizontalAdapter() : RecyclerView.Adapter<ListMovieHorizontalAdapter.ViewHolder>() {

    var listMovie: List<Movie>
        get() = listMovie
        set(value) {
            listMovie = value
            notifyDataSetChanged()
        }

    lateinit var context: Context
    lateinit var clickItemListener: ClickItemListener

    constructor(
        listMovie: List<Movie>,
        context: Context,
        clickItemListener: ClickItemListener
    ) : this() {
        this.listMovie = listMovie
        this.context = context
        this.clickItemListener = clickItemListener
    }


    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView!!) {
        @BindView(R.id.image_movie)
        var movieImageView: ImageView? = null

        @BindView(R.id.tv_movie_name)
        var movieNameTextView: TextView? = null

        fun binData(item: Movie, clickItemListener : ClickItemListener) {
            ButterKnife.bind(this, itemView)
            val urlImage: String = KEY_IMAGE + item.posterPath
            Glide.with(movieImageView!!)
                .load(urlImage)
                .into(movieImageView!!)
            movieNameTextView?.setText(item.title)
            itemView.setOnClickListener(View.OnClickListener {
                if (clickItemListener != null) {
                    clickItemListener.onClickMovie(adapterPosition)
                }
            })
        }

    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(LayoutInflater.from(context).inflate(R.layout.item_movie, parent, false))
    }

    override fun getItemCount(): Int {
        return listMovie.size
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        var item = listMovie.get(position)
        holder.binData(item, clickItemListener)
    }

    interface ClickItemListener {
        fun onClickMovie(position: Int)
    }
}